node/core/dispute-coordinator/src/scraping/mod.rs

```diff
@@ -69,9 +69,68 @@ impl ScrapedUpdates {
 	}
 }
 
+/// A structure meant to facilitate chain reversions in the event of a dispute
+/// concluding against a candidate. Each candidate hash maps to a vector of block
+/// number + hash pairs for all blocks which included the candidate. The entries
+/// in each vector are ordered by decreasing parent block number to facilitate
+/// minimal cost pruning.
+pub struct Inclusions {
+	inclusions_inner: BTreeMap<CandidateHash, BTreeMap<BlockNumber, Vec<Hash>>>,
+}
+
+impl Inclusions {
+	pub fn new() -> Self {
+		Self { inclusions_inner: BTreeMap::new() }
+	}
+
+	// Insert parent block into vector for the candidate hash it is including,
+	// maintaining ordering by decreasing parent block number.
+	pub fn insert(
+		&mut self,
+		candidate_hash: CandidateHash,
+		block_number: BlockNumber,
+		block_hash: Hash,
+	) {
+		if let Some(blocks_including) = self.inclusions_inner.get_mut(&candidate_hash) {
+			if let Some(blocks_at_height) = blocks_including.get_mut(&block_number) {
+				blocks_at_height.push(block_hash);
+			} else {
+				blocks_including.insert(block_number, Vec::from([block_hash]));
+			}
+		} else {
+			let mut blocks_including: BTreeMap<BlockNumber, Vec<Hash>> = BTreeMap::new();
+			blocks_including.insert(block_number, Vec::from([block_hash]));
+			self.inclusions_inner
+				.insert(candidate_hash, blocks_including);
+		}
+	}
+
+	pub fn remove_up_to_height(&mut self, height: &BlockNumber) {
+		for blocks_including in self.inclusions_inner.values_mut() {
+			*blocks_including = blocks_including.split_off(height);
```

Comment:
**nit:** This line gave me pause to look up `split_off` and think about whether this works. It also relies on the assumption that elements are stored in ascending order. Could just be me being slow today -- but I would add this:

```suggestion
            // Returns everything after the given key, including the key. This works because the blocks are sorted in ascending order.
			*blocks_including = blocks_including.split_off(height);
```